Throw IdentityException if operand of monent is a value type object

hzongaro opened the pull request:
Previously, specifying an instance of a value type class on a `monitorenter` bytecode would result in an `IllegalMonitorStateException`; a `java.lang.IdentityException` must now be thrown instead.

This commit introduces a non-helper, `<isIdentityObject>`, that checks that the operand of a `TR::monent` instruction is an instance of an identity class rather than a value type class.  If it is a value type class, it throws an `IdentityException`.  The actual test is expanded to test the flags field of the object's class during the Tree Lowering optimization pass.

The result of `<isIdentityObject`> is tested on a `ZEROCHK` operation that is generated before the `monent` during IL generation.

This change includes only required functional changes that allow JIT-compiled code to detect whether it should throw an `IdentityException`.  A planned improvement to Value Propagation will fold calls to `<isIdentityObject>` if the argument is recognized to be an instance of a value type class or recognized to be an instance of an identity type class at compile-time.

This change depends on eclipse-omr/omr#7674

Fixes #20984

hzongaro commented:
I have marked this pull request as draft, as it depends on eclipse-omr/omr#7674.  However, I believe it is complete and ready for review.

@a7ehuo, @0xdaryl, may I ask you to review this change?
@theresa-m, @hangshao0, may I ask you to review the VM-related portions of this change?

hangshao0 commented on runtime/codert_vm/cnathelp.cpp:
```diff
@@ -2731,6 +2731,20 @@ old_slow_jitThrowArrayIndexOutOfBounds(J9VMThread *currentThread)
+	OLD_JIT_HELPER_PROLOGUE(0);
+	buildJITResolveFrameForRuntimeCheck(currentThread);
+	exception = setCurrentExceptionFromJIT(currentThread, J9VMCONSTANTPOOL_JAVALANGIDENTITYEXCEPTION, NULL);
+#else /* defined(J9VM_OPT_VALHALLA_FLATTENABLE_VALUE_TYPES) */
```
The comment here should be /* defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */

hzongaro replied:
Corrected in cfe2348588b959fc64f648aba0238aea645c7e34

a7ehuo commented on runtime/compiler/compile/J9SymbolReferenceTable.hpp:
```diff
@@ -415,6 +427,18 @@ class SymbolReferenceTable : public OMR::SymbolReferenceTableConnector
+   /**
+    * \brief
+    *    Finds the <isIdentityObject> "non-helper" symbol reference, creating it if
+    *    necessary.  The non-helper is used to test whether an object is an instance
```
An extra space in `necessary.  The`

hzongaro replied:
That was intentional.  I always use two spaces after the period at the end of sentence, or after a colon.  It's a style that was common using typewriters that printed using monospace fonts like Courier, and as we usually look at our code in Courier font, I prefer to continue to follow that convention.

a7ehuo commented on runtime/compiler/runtime/asmprotos.h:
```diff
@@ -143,6 +143,9 @@ JIT_HELPER(jitThrowCurrentException);  // asm calling-convention helper
 JIT_HELPER(jitThrowException);  // asm calling-convention helper
 JIT_HELPER(jitThrowUnreportedException);  // asm calling-convention helper
 JIT_HELPER(jitThrowExceptionInInitializerError);  // asm calling-convention helper
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+JIT_HELPER(jitThrowIdentityException);  // asm calling-convention helper
+#endif /* defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
```
Same comment as the above regarding to J9VM_OPT_VALHALLA_VALUE_TYPES.

hzongaro replied:
Same as the previous comment - I think I had left this `ifdef` in because of the way I defined `old_slow_jitThrowIdentityException` in an earlier version of the code.  I will remove it.

hzongaro replied:
Corrected in commit 18dd4ab4d51a30c7e44bac5052fa00b535cec694

a7ehuo reviewed (commented):
Thank you very much for the fix! Just some minor comments.

hzongaro commented on runtime/compiler/runtime/Runtime.cpp:
```diff
@@ -1081,6 +1081,10 @@ void initializeCodeRuntimeHelperTable(J9JITConfig *jitConfig, char isSMP)
    SET(TR_typeCheckArrayStore,        (void *)jitTypeCheckArrayStoreWithNullCheck,   TR_Helper);
 #endif
 
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+   SET(TR_identityException,          (void *)jitThrowIdentityException, TR_Helper);
+#endif /* defined(J9VM_OPT_VALHALLA_VALUE_TYPES) */
```
I think at one point, I had the entire definition of `old_slow_jitThrowIdentityException` enclosed in an

```
#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
```

and so I conditionally compiled these uses of `jitThrowIdentityException`.  But then I ended up with link errors on a couple of platforms when building without valhalla enabled, so I adjusted the definition of `old_slow_jitThrowIdentityException` to allow it to compile even without value types enabled.  But I forgot to remove these `ifdefs`.

Yes, I think they can safely be removed.  Thanks for pointing that out!

hzongaro replied:
Corrected in commit 18dd4ab4d51a30c7e44bac5052fa00b535cec694

hzongaro commented on runtime/compiler/optimizer/TreeLowering.cpp:
```diff
@@ -1536,6 +1536,60 @@ StoreArrayElementTransformer::lower(TR::Node* const node, TR::TreeTop* const tt)
+   n97n        aloadi  <vft-symbol>
+   n77n          aload x
+   n96n    iconst 0x80000       // Test whether class is an identity class
+ * @endverbatim
```
Corrected in commit 18dd4ab4d51a30c7e44bac5052fa00b535cec694

hzongaro commented:
I addressed the review comments from @a7ehuo in commit Corrected in commit 18dd4ab4d51a30c7e44bac5052fa00b535cec694.  Once everyone is happy with these changes, I will squash the commits.

a7ehuo reviewed (approved):
LGTM. Thanks for the update!

hzongaro commented:
Squashed the commits.  Marking this change as ready for review now that eclipse-omr/omr#7674 has been promoted.

0xdaryl commented:
Jenkins test sanity xlinuxval,xmacval,,winval,plinuxval,aixval,alinuxval,zlinuxval,osxval jdknext

hzongaro commented:
Most test failures occurred because `sanity` is expanded to `sanity.functional,sanity.openjdk`.  See my draft pull request attempting to fix that problem:  #20316

The [build failure for x86-64 Windows](https://openj9-jenkins.osuosl.org/job/Build_JDKnext_x86-64_windows_valhalla_Personal/45) appears to be a build compiler problem:

```
01:11:58  100 warnings
01:14:11  /cygdrive/f/Users/jenkins/workspace/Build_JDKnext_x86-64_windows_valhalla_Personal/make/scripts/fixpath.sh: line 435: 55831 Segmentation fault      "$command" "${collected_args[@]}"
01:14:11  f:\Users\jenkins\workspace\Build_JDKnext_x86-64_windows_valhalla_Personal\test\hotspot\jtreg\serviceability\jvmti\valhalla\FieldAccessModify\libFieldAccessModify.cpp(237): fatal error C1001: Internal compiler error.
01:14:11  (compiler file 'msc1.cpp', line 1693)
01:14:11   To work around this problem, try simplifying or changing the program near the locations listed above.
```

The [x86-64 Linux build problem](https://openj9-jenkins.osuosl.org/job/Build_JDKnext_x86-64_linux_valhalla_Personal/134) appears to be infrastructure-related.


Jenkins test sanity.functional xlinuxval,winval jdknext

hzongaro commented:
The internal compiler error reported in the [`x86-64 Windows` build](https://openj9-jenkins.osuosl.org/job/Build_JDKnext_x86-64_windows_valhalla_Personal/46/) does not appear to be new.  I see the same problem in a [build for pull request](https://openj9-jenkins.osuosl.org/job/Build_JDKnext_x86-64_windows_valhalla_Personal/44/) #20456 from 2024-10-31.
